statsmodels/kernel_methods/kde.py

```diff
@@ -0,0 +1,307 @@
+        self._kernel = None
+
+        self.exog = exog
+
+        for n in kwords:
+            if hasattr(self, n):
+                setattr(self, n, kwords[n])
+            else:
+                raise AttributeError("Error, unknown attribute: '{}'".format(n))
+
+        if self._method is None:
+            self._method = default_method()
+
+    def copy(self):
+        """
+        Shallow copy of the KDE object
+        """
+        res = shallow_copy(self)
+        res._method = self._method.copy()
+        res._axis_type = shallow_copy(self._axis_type)
+        res._lower = shallow_copy(self._lower)
+        res._upper = shallow_copy(self._upper)
+        return res
+
+    @property
+    def lower(self):
+        """
+        List with the lower bound for the domain on each dimension. None for automatic computation of the bound.
+        """
+        return self._lower
+
+    @lower.setter
+    def lower(self, val):
+        self._lower = val
+
+    @lower.deleter
+    def lower(self):
+        self._lower = -np.inf*np.ones(self.ndim, dtype=float)
+
+    @property
+    def upper(self):
+        """
+        List with the upper bound for the domain on each dimension. None for automatic computation of the bound.
+        """
+        return self._upper
+
+    @upper.setter
+    def upper(self, val):
+        self._upper = val
+
+    @upper.deleter
+    def upper(self):
+        self._upper = np.inf*np.ones(self.ndim, dtype=float)
+
+    @property
+    def exog(self):
+        """
+        2D array with exogenous data. The array has shape NxD for N points in D dimension.
+        """
+        return self._exog
+
+    @exog.setter
+    def exog(self, value):
+        value = atleast_2df(value).astype(float)
+        ndim = value.shape[1]
+        if ndim != self.ndim:
+            self._axis_type = AxesType('C' * ndim)
+            if ndim == 1:
+                self._lower = -np.inf
+                self._upper = np.inf
+            else:
+                self._lower = [-np.inf] * ndim
+                self._upper = [np.inf] * ndim
+        self._exog = value
+
+    @property
+    def ndim(self):
+        """
+        Number of dimensions of the problem.
+        """
+        if self._exog is None:
+            return 0
+        return self._exog.shape[1]
+
+    @property
+    def npts(self):
+        """
+        Number of points in the problem.
+        """
+        return self._exog.shape[0]
+
+    @property
+    def axis_type(self):
+        """
+        Set the types of Axis. Defined in :py:class:`AxesTypes`
+        """
+        return self._axis_type
+
+    @axis_type.setter
+    def axis_type(self, value):
+        self._axis_type = value
+
+    @axis_type.deleter
+    def axis_type(self):
+        self._axis_type = AxesType()
+
+    @property
+    def method(self):
+        """
+        Method used to estimate the KDE.
+        """
+        return self._method
+
+    @method.setter
+    def method(self, m):
```

Comment:
For `_kde1d_methods.py`, I can split it so there is one file per method.

For the `method` property, the idea is to provide a unique point of entry to define the KDE problem, and be able to choose the fitting method independently. For example, if you have an existing KDE object `k` with the default fitting method, you may want to try using the `LineageCombination` method, which is a first order approximation for bounded domains. You would do:

    k.method = kde_methods.LineageCombination()
    mod = k.fit()

As for the `fit` method, it is defined in the various classes imported in `kde_methods`. I have put below a more precise description of the role of this `fit` method.